packages/neuron-wallet/src/block-sync-renderer/sync/connector.ts

```diff
@@ -41,15 +49,147 @@ export interface AppendScript {
+    this.processNextBlockNumberQueue.error((err: any) => {
+      logger.error(`Error in processing next block number queue: ${err}`)
+    })
+
+    this.indexerQueryQueue = queue(async (query: any) => {
+      return await this.collectLiveCellsByScript(query)
+    })
+  }
+
+  public stop(): void {
+    this.pollingIndexer = false
+  }
+
+  protected async processNextBlockNumber() {
+    this.processNextBlockNumberQueue?.push(null)
+    await this.processNextBlockNumberQueue?.drain()
+  }
+
+  protected async getTxHashesWithNextUnprocessedBlockNumber(): Promise<[string | undefined, string[]]> {
+    const txHashCachesByNextBlockNumberAndAddress = await Promise.all(
+      [...this.addressesByWalletId.entries()].map(async ([walletId]) =>
+        IndexerCacheService.nextUnprocessedTxsGroupedByBlockNumber(walletId)
+      )
+    )
+    const groupedTxHashCaches = txHashCachesByNextBlockNumberAndAddress.flat().reduce((grouped, txHashCache) => {
+      if (!grouped.get(txHashCache.blockNumber.toString())) {
+        grouped.set(txHashCache.blockNumber.toString(), [])
+      }
+      grouped.get(txHashCache.blockNumber.toString())!.push(txHashCache)
+
+      return grouped
+    }, new Map<string, Array<IndexerTxHashCache>>())
+
+    const nextUnprocessedBlockNumber = [...groupedTxHashCaches.keys()].sort((a, b) => parseInt(a) - parseInt(b)).shift()
+
+    if (!nextUnprocessedBlockNumber) {
+      return [undefined, []]
+    }
+
+    const txHashCachesInNextUnprocessedBlockNumber = groupedTxHashCaches.get(nextUnprocessedBlockNumber)
+
+    return [nextUnprocessedBlockNumber, txHashCachesInNextUnprocessedBlockNumber!.map(({ txHash }) => txHash)]
+  }
+
+  protected async notifyAndSyncNext(indexerTipNumber: number) {
+    const nextUnprocessedBlockNumber = await IndexerCacheService.nextUnprocessedBlock([
+      ...this.addressesByWalletId.keys(),
+    ])
+    if (nextUnprocessedBlockNumber) {
+      this.blockTipsSubject.next({
+        cacheTipNumber: parseInt(nextUnprocessedBlockNumber),
+        indexerTipNumber,
+      })
+      if (!this.processingBlockNumber) {
+        await this.processNextBlockNumber()
+      }
+      return true
+    }
+    this.blockTipsSubject.next({
+      cacheTipNumber: indexerTipNumber,
+      indexerTipNumber,
+    })
+    return false
+  }
+
+  public async getLiveCellsByScript(query: LumosCellQuery) {
+    return new Promise((resolve, reject) => {
+      this.indexerQueryQueue!.push(query, (err: any, result: unknown) => {
+        if (err) {
+          return reject(err)
+        }
+        resolve(result)
+      })
+    })
+  }
+
+  private async collectLiveCellsByScript(query: LumosCellQuery) {
+    const { lock, type, data } = query
+    if (!lock && !type) {
+      throw new Error('at least one parameter is required')
+    }
+
+    const queries: QueryOptions = {}
+    if (lock) {
+      queries.lock = lock
+    }
+    if (type) {
+      queries.type = type
+    }
+    queries.data = data || 'any'
```

Comment:
Why not set `lock`, `type` and `data` when `queries` is initialized

Reply:
Here codes move from https://github.com/nervosnetwork/neuron/blob/develop/packages/neuron-wallet/src/block-sync-renderer/sync/indexer-connector.ts#L113-L145. Because of the `light-connector` and `indexer-connector` can share this function.

Reply:
It seems that the `initialize and then assign` logic was left to minimize code change when it was refactored with lumos @homura